Support contextual serializers

I can do this:

```
@Serializable
data class Foo(
  @Contextual
  val d: LocalDate
)
```

and create a JSON serializer like this:

```
Json {
    serializersModule = SerializersModule {
        contextual(LocalDateAsLongSerializable)
    }
}
```

however if I try to do a similar same thing with Avro4k like this:

```
Avro(SerializersModule {
    contextual(LocalDateSerializer())
})
```

at runtime I get the error:

```
Caused by: kotlinx.serialization.SerializationException: Unsupported type kotlinx.serialization.ContextualSerializer<LocalDate> of CONTEXTUAL
	at com.github.avrokotlin.avro4k.schema.SchemaForKt.schemaFor(SchemaFor.kt:180)
	at com.github.avrokotlin.avro4k.schema.ClassSchemaFor.buildField(ClassSchemaFor.kt:78)
	at com.github.avrokotlin.avro4k.schema.ClassSchemaFor.dataClassSchema(ClassSchemaFor.kt:63)
	at com.github.avrokotlin.avro4k.schema.ClassSchemaFor.schema(ClassSchemaFor.kt:43)
	at com.github.avrokotlin.avro4k.Avro.schema(Avro.kt:266)
	at com.github.avrokotlin.avro4k.Avro.schema(Avro.kt:269)
	at com.github.avrokotlin.avro4k.AvroOutputStreamBuilder.to(Avro.kt:121)
```

--- a/src/main/kotlin/com/github/avrokotlin/avro4k/schema/SchemaFor.kt
+++ b/src/main/kotlin/com/github/avrokotlin/avro4k/schema/SchemaFor.kt
@@ -4,9 +4,11 @@ import com.github.avrokotlin.avro4k.AnnotationExtractor
 import com.github.avrokotlin.avro4k.Avro
 import com.github.avrokotlin.avro4k.RecordNaming
 import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.InternalSerializationApi
 import kotlinx.serialization.SerializationException
 import kotlinx.serialization.descriptors.*
 import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.serializerOrNull
 import org.apache.avro.Schema
 import org.apache.avro.SchemaBuilder
 
@@ -155,6 +157,7 @@ class NullableSchemaFor(private val schemaFor: SchemaFor, private val annotation
    }
 }
 
+@OptIn(InternalSerializationApi::class)
 @ExperimentalSerializationApi
 fun schemaFor(serializersModule: SerializersModule,
               descriptor: SerialDescriptor,
@@ -181,6 +184,18 @@ fun schemaFor(serializersModule: SerializersModule,
          PrimitiveKind.FLOAT -> SchemaFor.FloatSchemaFor
          PrimitiveKind.BOOLEAN -> SchemaFor.BooleanSchemaFor
          SerialKind.ENUM -> EnumSchemaFor(descriptor)
+         SerialKind.CONTEXTUAL -> schemaFor(
+            serializersModule,
+            requireNotNull(
+               serializersModule.getContextualDescriptor(descriptor)
+                  ?: descriptor.capturedKClass?.serializerOrNull()?.descriptor
+            ) {
+               "Contextual or default serializer not found for $descriptor "
+            },
+            annos,
+            namingStrategy,
+            resolvedSchemas
+         )
          PolymorphicKind.SEALED -> SealedClassSchemaFor(descriptor, namingStrategy, serializersModule, resolvedSchemas)
          StructureKind.CLASS, StructureKind.OBJECT -> when (descriptor.serialName) {
             "kotlin.Pair" -> PairSchemaFor(descriptor, namingStrategy, serializersModule, resolvedSchemas)

--- a/src/test/kotlin/com/github/avrokotlin/avro4k/schema/ContextualSchemaTest.kt
+++ b/src/test/kotlin/com/github/avrokotlin/avro4k/schema/ContextualSchemaTest.kt
@@ -1,0 +1,51 @@
+package com.github.avrokotlin.avro4k.schema
+
+import com.github.avrokotlin.avro4k.Avro
+import com.github.avrokotlin.avro4k.serializer.InstantSerializer
+import com.github.avrokotlin.avro4k.serializer.InstantToMicroSerializer
+import io.kotest.assertions.throwables.shouldThrow
+import io.kotest.core.spec.style.StringSpec
+import io.kotest.matchers.shouldBe
+import kotlinx.serialization.Contextual
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.modules.contextual
+import org.apache.avro.Schema
+import java.lang.IllegalArgumentException
+import java.time.Instant
+
+class ContextualSchemaTest : StringSpec({
+
+    "schema for contextual serializer" {
+        @Serializable
+        data class Test(
+            @Contextual
+            val ts: Instant,
+            @Contextual
+            val withFallback: Int?,
+        )
+
+        val format1 = Avro(
+            serializersModule = SerializersModule {
+                contextual(InstantSerializer())
+            }
+        )
+
+        val format2 = Avro(
+            serializersModule = SerializersModule {
+                contextual(InstantToMicroSerializer())
+            }
+        )
+
+        val schema1 = format1.schema(Test.serializer())
+        val schema2 = format2.schema(Test.serializer())
+
+        shouldThrow<IllegalArgumentException> {
+            Avro.default.schema(Test.serializer())
+        }
+        val expected1 = Schema.Parser().parse(javaClass.getResourceAsStream("/contextual_1.json"))
+        schema1 shouldBe expected1
+        val expected2 = Schema.Parser().parse(javaClass.getResourceAsStream("/contextual_2.json"))
+        schema2 shouldBe expected2
+    }
+})

--- a/src/test/resources/contextual_1.json
+++ b/src/test/resources/contextual_1.json
@@ -1,0 +1,21 @@
+{
+  "type": "record",
+  "name": "Test",
+  "namespace": "com.github.avrokotlin.avro4k.schema.ContextualSchemaTest",
+  "fields": [
+    {
+      "name": "ts",
+      "type": {
+        "type": "long",
+        "logicalType": "timestamp-millis"
+      }
+    },
+    {
+      "name": "withFallback",
+      "type": [
+        "null",
+        "int"
+      ]
+    }
+  ]
+}

--- a/src/test/resources/contextual_2.json
+++ b/src/test/resources/contextual_2.json
@@ -1,0 +1,21 @@
+{
+  "type": "record",
+  "name": "Test",
+  "namespace": "com.github.avrokotlin.avro4k.schema.ContextualSchemaTest",
+  "fields": [
+    {
+      "name": "ts",
+      "type": {
+        "type": "long",
+        "logicalType": "timestamp-micros"
+      }
+    },
+    {
+      "name": "withFallback",
+      "type": [
+        "null",
+        "int"
+      ]
+    }
+  ]
+}

